### PR TITLE
[FLINK-3554] [streaming] Emit a MAX Watermark after finite sources finished

### PR DESCRIPTION
**Pull request by StephanEwen**
Finite sources now emit a MAX_LONG watermark once they cleanly finished.

This prevents loss of pending windows over finite sources when they finish.

Canceling and stopping sources will NOT emit a MAX Watermark.

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StoppableStreamSource.java [hunk 1]
```diff
@@ -41,7 +41,14 @@ public StoppableStreamSource(SRC sourceFunction) {
 		super(sourceFunction);
 	}
 
+	/**
+	 * Marks the source a stopped and calls {@link StoppableFunction#stop()} on the user function.
+	 */
 	public void stop() {
+		// important: marking the source as stopped has to happen before the function is stopped.
+		// the flag that tracks this status is volatile, to the memory model also guarantees
```
Comment (tillrohrmann) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StoppableStreamSource.java]: typo in comment: to => so

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamSource.java [hunk 2]
```diff
@@ -66,19 +71,60 @@ public void run(final Object lockingObject, final Output<StreamRecord<OUT>> coll
...
+			}
+		} finally {
+			// make sure that the context is closed in any case
+			ctx.close();
+		}
 	}
 
 	public void cancel() {
+		// important: marking the source as stopped has to happen before the function is stopped.
+		// the flag that tracks this status is volatile, to the memory model also guarantees
```
Comment (tillrohrmann) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamSource.java]: typo in comment: to => so

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/SourceFunctionTest.java [hunk 3]
```diff
@@ -61,9 +61,9 @@ public void generateSequenceTest() throws Exception {
 		assertEquals(expectedList, actualList);
 	}
 
-	@Test
-	public void socketTextStreamTest() throws Exception {
-		// TODO: does not work because we cannot set the internal socket anymore
+// TODO: does not work because we cannot set the internal socket anymore
+//	@Test
+//	public void socketTextStreamTest() throws Exception {
```
Comment (tillrohrmann) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/SourceFunctionTest.java]: Why not removing this test?

**Comment by tillrohrmann** [comment]
Changes look good to me. Really good fix @StephanEwen :-) +1 for merging.

**Comment by aljoscha** [comment]
+1

**Comment by rmetzger** [comment]
Looks like the tests are failing

**Comment by StephanEwen** [comment]
Adjusting the tests now. Some were actually assuming the old (wrong) behavior.

**Comment by wenlonglwl** [comment]
WaterMark seems do not work well with count window ?

**Comment by aljoscha** [comment]
Hi,
what are you referring to? The watermark does not interact with count windows.

**Comment by wenlonglwl** [comment]
just want to remind, this dose not completely solve the problem of loss of pending windows over finite sources~ looking forward to the solutions~

**Comment by aljoscha** [comment]
Yes, you are right and we are thinking about it.

**Comment by StephanEwen** [comment]
It is a matter of how we want to define the semantics.

For windows that do not depend on event time, should incomplete windows be emitted or not? There are arguments for both. So far, most users requested to not have incomplete windows emitted.

**Comment by aljoscha** [comment]
Yes, right now we favor correctness over completeness. For example in a Count window, if we emitted an unfinished count window then the count of the elements in the window would not be what the user specified.

For Event-Time windows we are correct in emitting a window upon arrival of the final Long.MAX_VALUE watermark.

**Comment by StephanEwen** [comment]
So, the semantics are that at the end of a finite source, the "end of event time" is reached, which would make the emission of pending event time semantically correct.

So far there is nothing equivalent for processing time.

**Comment by wenlonglwl** [comment]
In our production environment, window is used to make a small batch, for some operator that should be done in batch to improve the throughput, such as writing data to HBase, whether the count reach or not, is not significant, but no losing data is. I think this is a common user story.

**Comment by aljoscha** [comment]
Yes, that's  a problem. Could you maybe open an issue for that here: https://issues.apache.org/jira/browse/FLINK

In the meantime, I have a workaround for you. You can use a modified version of the CountTrigger:

```
public class ClosingCountTrigger<W extends Window> extends Trigger<Object, W> {
	private static final long serialVersionUID = 1L;

	private final long maxCount;

	private final ValueStateDescriptor<Long> stateDesc =
			new ValueStateDescriptor<>("count", LongSerializer.INSTANCE, 0L);


	private ClosingCountTrigger(long maxCount) {
		this.maxCount = maxCount;
	}

	@Override
	public TriggerResult onElement(Object element, long timestamp, W window, TriggerContext ctx) throws IOException {
		ValueState<Long> count = ctx.getPartitionedState(stateDesc);
		long currentCount = count.value() + 1;
		count.update(currentCount);
		if (currentCount >= maxCount) {
			count.update(0L);
			return TriggerResult.FIRE_AND_PURGE;
		}
		
		// register timer for final watermark
		ctx.registerEventTimeTimer(Long.MAX_VALUE);
		return TriggerResult.CONTINUE;
	}

	@Override
	public TriggerResult onEventTime(long time, W window, TriggerContext ctx) {
		// final triggering
		return TriggerResult.FIRE_AND_PURGE;
	}

	@Override
	public TriggerResult onProcessingTime(long time, W window, TriggerContext ctx) throws Exception {
		return TriggerResult.CONTINUE;
	}

	@Override
	public void clear(W window, TriggerContext ctx) throws Exception {
		ctx.getPartitionedState(stateDesc).clear();
		ctx.deleteEventTimeTimer(Long.MAX_VALUE);
	}

	@Override
	public String toString() {
		return "ClosingCountTrigger(" +  maxCount + ")";
	}

	/**
	 * Creates a trigger that fires once the number of elements in a pane reaches the given count.
	 *
	 * @param maxCount The count of elements at which to fire.
	 * @param <W> The type of {@link Window Windows} on which this trigger can operate.
	 */
	public static <W extends Window> ClosingCountTrigger<W> of(long maxCount) {
		return new ClosingCountTrigger<>(maxCount);
	}
}
```

you can use it like this:

```
stream
    .keyBy(...)
    .window(GlobalWindows.create())
    .trigger(ClosingCountTrigger.of(<my count>)
    .apply() // or .reduce()
```

This will behave like the normal count trigger but also register a callback for the final `Long.MAX_VALUE` watermark. Then, when the final watermark arrives it will trigger one last window emission.

**Comment by wenlonglwl** [comment]
got it~ 3q @aljoscha

**Pull request closed without merging**